gitutil.py

```diff
@@ -0,0 +1,137 @@
+
+    Returns
+    ------
+    str
+        The hexadecimal SHA-1 hash of the created commit
+    '''
+    ensure_directory_exists(repo_path)
+    repo_path = os.path.abspath(repo_path)
+
+    ensure_not_empty(submodule_path)
+    if '/' in submodule_path:
+        fail('Specified submodule-path is not in the top-level of the repository: {s}'.format(s=submodule_path))
+
+    ensure_not_empty(tree_ish)
+    ensure_not_empty(commit_hash)
+    ensure_not_empty(author)
+    ensure_not_empty(email)
+
+    repo = git.Repo(repo_path)
+
+    # Fetch tree-ish from git.repo and create mktree-parseable string-representation of it.
+    # Since submodule-objects contain the SHA of the to-be-checked-out version of the
+    # target repository, we replace that during the string building.
+    tree = repo.tree(tree_ish)
+    tree_representation = '\n'.join(_serialise_and_patch_tree(tree, submodule_path, commit_hash))
+
+    # Pass the serialised tree to git mk-tree using GitPython. We cannot do this in GitPython
+    # directly as it does not support arbitrary tree manipulation.
+    auto_interrupt = repo.git.mktree(istream = subprocess.PIPE, as_process=True)
+    process = auto_interrupt.proc
+    stdout, _= process.communicate(input=tree_representation.encode())
+
+    # returned string is byte-encoded and newline-terminated
+    new_sha = stdout.decode('utf-8').strip()
+
+    # Create a new commit in the repo's object database from the newly created tree.
+    actor = git.Actor(author, email)
+    parent_commit = repo.commit(tree_ish)
+    commit = git.Commit.create_from_tree(
+      repo = repo,
+      tree = new_sha,
+      parent_commits = [parent_commit],
+      message='Upgrade submodule {s} to commit {c}'.format(
+          s=submodule_path,
+          c=commit_hash,
+      ),
+      author=actor,
+      committer=actor,
+    )
+
+    return commit.hexsha
+
+
+def _serialise_and_patch_tree(
+    tree: git.Tree,
+    submodule_path: str,
+    commit_hash: str,
+):
+    '''Create a generator that yields the string-representation of the passed tree as understood by
+    git ls-tree, one line at a time. Replaces the commit-hash of the tree element with the given one
+    if it is a submodule and its submodule path matches the passed `submodule_path`.
+    '''
+    # Since there is no direct way to get the ls-tree representation from GitPython we need to gather
+    # the necessary information for each element of the given tree ourselves.
+    submodule_found = False
+    for tree_element in tree:
+        # GitPython uses the special type 'submodule' for submodules whereas git uses 'commit'.
+        if tree_element.type == 'submodule':
+            element_type = 'commit'
+            # Replace the hash the of the 'commit'-tree with the passed value if the submodule is at the specified path
+            if tree_element.path == submodule_path:
+                element_sha = commit_hash
+                submodule_found = True
```

Comment:
might be discarded again, as you'll already check for the existence of a sub-module at the given path earlier

Reply:
since you are no longer evaluating it, this should be removed (as in: must)